### PR TITLE
feat: Implement non-partitioned tmp table

### PR DESCRIPTION
**Pull request by svdimchenko**
### Description

Resolves: https://github.com/dbt-athena/dbt-athena/issues/396

Currently we have partitioned temp table creation for table and increment materialisation.
However this can lead to over price and over resources consumption as model's sql can be complicated. So that batches split run on this complicated sql may be not the very optimal way.

This PR implements the logic when we skip partitioning config while creating new tmp table and then we calculate batches and run insert statement based on this non-partitioned table.

## Models used to test
table iceberg
```
{{
  config(
    schema='test_schema',
    materialized='table',
    partitioned_by=['DAY(date_column)', 'doy'],
    table_type='iceberg'
  )
}}

SELECT
    CAST(date_column AS DATE) as date_column,
    doy(date_column) as doy,
    random() as rnd
FROM (
    VALUES (
        SEQUENCE(FROM_ISO8601_DATE('2020-01-01'), FROM_ISO8601_DATE('2023-07-24'), INTERVAL '1' DAY)
    )
) AS t1(date_array)
CROSS JOIN UNNEST(date_array) AS t2(date_column)
```
increment iceberg
```
{{
  config(
    schema='test_schema',
    materialized='incremental',
    incremental_strategy='merge',
    partitioned_by=['DAY(date_column)', 'doy'],
    table_type='iceberg',
    unique_key=['doy']
  )
}}

select * from {{ ref('test_dbt_table') }}
```

table hive
```
{{
  config(
    schema='test_schema',
    materialized='table',
    partitioned_by=['date_column', 'doy'],
    table_type='hive'
  )
}}

SELECT
    random() as rnd,
    CAST(date_column AS DATE) as date_column,
    doy(date_column) as doy
FROM (
    VALUES (
        SEQUENCE(FROM_ISO8601_DATE('2020-01-01'), FROM_ISO8601_DATE('2023-07-24'), INTERVAL '1' DAY)
    )
) AS t1(date_array)
CROSS JOIN UNNEST(date_array) AS t2(date_column)
```

## Checklist
- [x] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You kept your Pull Request small and focused on a single feature or bug fix.
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
     {%- set partitioned_keys = adapter.format_partition_keys(partitioned_by) -%}
     {% do log('PARTITIONED KEYS: ' ~ partitioned_keys) %}
 
     {% call statement('get_partitions', fetch_result=True) %}
-        select distinct {{ partitioned_keys }} from ({{ sql }}) order by {{ partitioned_keys }};
+        {%- if as_subquery -%}
+            select distinct {{ partitioned_keys }} from ({{ sql }}) order by {{ partitioned_keys }};
+        {%- else -%}
+            select distinct {{ partitioned_keys }} from {{ sql }} order by {{ partitioned_keys }};
+        {%- endif -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql:12]: I'm not fully getting why we need `as_subquery`? do we have any cases where we should use as_subquery=False?
or you implemented this way to make this macro more future proof?

File: dbt/include/athena/macros/materializations/models/table/create_table_as.sql [hunk 2]
```diff
@@ -90,41 +90,70 @@
 
 {% macro create_table_as_with_partitions(temporary, relation, sql) -%}
 
-    {% set partitions_batches = get_partition_batches(sql) %}
+    {%- set stg_relation = api.Relation.create(
+            identifier=relation.identifier ~ '__stg',
```
Comment (nicor88) [inline dbt/include/athena/macros/materializations/models/table/create_table_as.sql:94]: we generally call those relation `tmp_relation` see https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/table/table.sql#L21-L25 just for consistency of naming I will suggest to pick `tmp_relation` and eventually consider to use `__tmp` as prefix, or something like 
`__tmp_no_partitions` to make even more explicitly the context of that tmp table name

File: dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql (hunk 1 above)
Comment (svdimchenko) [reply dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql:12]: yeah, here we use it https://github.com/dbt-athena/dbt-athena/pull/405/files#diff-f2eabd411bb7e43d849bdc621ef530f1304f39e18c4f1dcf73a887af4085b0d8R109

Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql:12]: what I mean is that we use `as_subquery=False`, means we use the tmp table that is more efficient. What will be the case where we use with `as_subquery=True`?

File: dbt/include/athena/macros/materializations/models/table/create_table_as.sql (hunk 2 above)
Comment (svdimchenko) [reply dbt/include/athena/macros/materializations/models/table/create_table_as.sql]: set to `__tmp_not_partitioned`

File: dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql (hunk 1 above)
Comment (svdimchenko) [reply dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql:12]: ah, I just tried not to break previous logic. Mb someone uses that macro already

Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/helpers/get_partition_batches.sql:12]: ok thanks, we could leave it, as it makes it backcompatible.

**Comment by nicor88** [comment]
Great work 💯 - our functional tests should cover this implementation already I believe and they seems ✅

**Review by nicor88 — APPROVED** [review]

**Pull request merged**